unified-runtime/source/adapters/level_zero/usm.hpp

```diff
@@ -11,25 +11,83 @@
 
 #include "common.hpp"
 
+#include "ur_api.h"
 #include "ur_pool_manager.hpp"
+#include <set>
 #include <umf_helpers.hpp>
 
 usm::DisjointPoolAllConfigs InitializeDisjointPoolConfig();
 
+class EnqueuedPool {
+public:
+  struct Allocation {
+    void *Ptr;
+    size_t Size;
+    ur_event_handle_t Event;
+    ur_queue_handle_t Queue;
+    size_t Alignment;
+  };
+
+  ~EnqueuedPool();
+  std::optional<Allocation> getBestFit(size_t Size, size_t Alignment,
+                                       ur_queue_handle_t Queue);
+  void insert(void *Ptr, size_t Size, ur_event_handle_t Event,
+              ur_queue_handle_t Queue);
+  bool cleanup();
+  bool cleanupForQueue(ur_queue_handle_t Queue);
+
+private:
+  struct Comparator {
+    bool operator()(const Allocation &lhs, const Allocation &rhs) const {
+      if (lhs.Queue != rhs.Queue) {
+        return lhs.Queue < rhs.Queue; // Compare by queue handle first
+      }
+      if (lhs.Alignment != rhs.Alignment) {
+        return lhs.Alignment < rhs.Alignment; // Then by alignment
+      }
+      if (lhs.Size != rhs.Size) {
+        return lhs.Size < rhs.Size; // Then by size
+      }
+      return lhs.Ptr < rhs.Ptr; // Finally by pointer address
+    }
+  };
+
+  using AllocationSet = std::set<Allocation, Comparator>;
+  ur_mutex Mutex;
+  AllocationSet Freelist;
+};
+
+struct UsmPool {
```

Comment:
Would it be possible to store `AsyncPool` as optional member in pool_descriptor? By doing that we could have pool_descriptor stored as 'tag' in umf pools, making getPoolByHandle unnecessary. This is the approach I took here: https://github.com/intel/llvm/pull/17230

Reply:
I've tried to move it to pool descriptor as a template argument type, it requires so much changes to satisfy the compiler that I gave up. I could try to do that in other PR.

Reply:
Ok, we can do that later then.